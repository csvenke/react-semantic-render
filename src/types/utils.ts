import * as PropTypes from 'prop-types';
import * as React from 'react';

export type Map =
  | ((item?: any, index?: number, array?: any[]) => React.ReactNode)
  | null;

export type Render = (() => React.ReactNode) | null;

export type Children = React.ReactNode | (() => React.ReactNode) | null;

export interface IRenderProps {
  /** Shorthand for primary content. */
  render?: Render;

  /** Primary content. */
  children?: Children;
}

export const renderPropsPropTypes = {
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  render: PropTypes.func,
};
