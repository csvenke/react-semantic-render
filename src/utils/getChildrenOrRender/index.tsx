import * as React from 'react';

import { isEmptyChildren } from '../isEmptyChildren';

export const getChildrenOrRender = (children?: any, render?: () => React.ReactNode) => {
  if (typeof children === 'function') {
    return children();
  }

  if (children && !isEmptyChildren(children)) {
    return React.Children.only(children);
  }

  if (render) {
    return render();
  }

  return null;
};
