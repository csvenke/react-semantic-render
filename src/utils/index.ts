export * from './isEmptyChildren';
export * from './isPromise/isPromise';
