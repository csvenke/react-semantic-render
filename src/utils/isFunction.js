function isFunction(input) {
  return typeof input === "function";
}

export default isFunction;
