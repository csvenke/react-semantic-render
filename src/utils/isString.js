function isString(input) {
  return typeof input === "string";
}

export default isString;
