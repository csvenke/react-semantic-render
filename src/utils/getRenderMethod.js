import isFunction from './isFunction'

function getRenderMethod (children, render) {
  if (children) {
    return children
  }

  if (render && isFunction(render)) {
    return render
  }

  return null
}

export default getRenderMethod
