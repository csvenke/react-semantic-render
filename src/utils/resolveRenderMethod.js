import React from "react";

import isEmptyChildren from "./isEmptyChildren";
import isFunction from "./isFunction";
import isString from "./isString";

function resolveRenderMethod(method) {
  if (isFunction(method)) {
    return method.call();
  }

  if (!isEmptyChildren(method) && React.isValidElement(method)) {
    return React.Children.only(method);
  }

  if (isString(method)) {
    return method;
  }

  return null;
}

export default resolveRenderMethod;
