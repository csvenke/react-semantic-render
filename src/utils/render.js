import getRenderMethod from "./getRenderMethod";
import resolveRenderMethod from "./resolveRenderMethod";

function render(props) {
  const method = getRenderMethod(props.children, props.render);
  return resolveRenderMethod(method);
}

export default render;
