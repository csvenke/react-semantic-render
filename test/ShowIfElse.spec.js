import React from 'react'
import renderer from 'react-test-renderer'

import { createTests } from './utils'
import { ShowIfElse as ShowIfElse1 } from '../dist/index.cjs'
import { ShowIfElse as ShowIfElse2 } from '../dist/index.esm'
import ShowIfElse3 from '../dist/ShowIfElse'
import ShowIfElse4 from '../src/ShowIfElse'

const runTests = createTests(ShowIfElse1, ShowIfElse2, ShowIfElse3, ShowIfElse4)

runTests(ShowIfElse => {
  const createComponent = condition => (
    <ShowIfElse
      condition={condition}
      if={() => <div>if</div>}
      else={() => <div>else</div>}
    />
  )

  test('should return content from if', () => {
    const component = renderer.create(createComponent(true))
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('should return content from else', () => {
    const component = renderer.create(createComponent(false))
    expect(component.toJSON()).toMatchSnapshot()
  })
})
