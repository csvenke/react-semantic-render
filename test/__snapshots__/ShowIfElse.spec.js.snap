// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return content from else 1`] = `
<div>
  else
</div>
`;

exports[`should return content from else 2`] = `
<div>
  else
</div>
`;

exports[`should return content from else 3`] = `
<div>
  else
</div>
`;

exports[`should return content from else 4`] = `
<div>
  else
</div>
`;

exports[`should return content from if 1`] = `
<div>
  if
</div>
`;

exports[`should return content from if 2`] = `
<div>
  if
</div>
`;

exports[`should return content from if 3`] = `
<div>
  if
</div>
`;

exports[`should return content from if 4`] = `
<div>
  if
</div>
`;
